// prestodb/test.js
'use strict';

const debug = require('debug')('node-prestodb:index.js');
const fetch = require('node-fetch');

// @TODO: move to configs
const prestoURL = 'http://profiler-reader-01.reddrummer.com:8889';
const prestoHeaders = { 'X-Presto-User': 'presto' };

function loopNextUri(uri) {
  debug(`loopNextUri('${uri}')`);
  return new Promise((resolve, reject) => {
    fetch(uri, { headers: prestoHeaders })
      .then((res) => res.json())
      .then((body) => {
        debug(`JSON body props: ${Object.keys(body)}`);

        // has columns

        // has data
        if (body.data) {
          debug('emit data', body.data);
          //, body.data.toString());
        }


        // has nextUri
        if (body.nextUri) {
          setTimeout(() => {
            loopNextUri(body.nextUri);
          }, 200);
        } else {
          resolve(body);
        }

      })
      .catch(reject);
  });
}

function sendStatement(query) {
  return new Promise((resolve, reject) => {
    fetch(`${prestoURL}/v1/statement`, {
      method: 'POST',
      body: query,
      headers: prestoHeaders
    }).then((res) => res.json())
      .then((body) => {
        debug('JSON body props: ', Object.keys(body));
        debug(`emit stats {${Object.keys(body.stats)}}`);
        resolve(loopNextUri(body.nextUri));
      })
      .catch(reject);
  });
}

let query = 'SELECT * FROM hive.vivo.linhas_tmp_201703 LIMIT 100';
sendStatement(query).then((result) => {
  console.log('query result:', result);
  //console.log(arguments);
}).catch((e) => {
  console.error('Bad');
  console.error(e);
});


/* TODO: use EventEmitter
const EventEmitter = require('events');// @see: https://nodejs.org/api/events.html
class Statement extends EventEmitter {
  constructor(client, query) {
    super();
    this.query = query;
    this.client = client;
    console.log(this);
  }
}

module.exports = Statement;
*/
